Core: Implement source-ids to deal with multi arguments transforms

jbonofre opened the pull request:
Following the update on the spec regarding `source-id` and `source-ids` (thanks again @Fokko 😄 ), here's the PR to introduce `source-ids` field in partition spec.

A few notes:
1. Internal representation is now based on `source-ids`
2. Serialization/deserialization supports `source-id` and `source-ids` elements in the json (exclusive), both populating `source-ids` internal representation (as `List<Integer>`)
3. The `TestPartitionSpecParser` is still testing `source-id`, but also `source-ids` parsing, and neither `source-id` and `source-ids` presence (throwing `IllegalArgumentException` in that case)
4. For backward compatibility (especially for the transforms), source id is still supported (using the first element in the internal representation) and some methods have been flagged as deprecated to encourage use of source ids.

jbonofre commented:
@rdblue @Fokko @RussellSpitzer ^^ Thanks !

ajantha-bhat commented on api/src/main/java/org/apache/iceberg/PartitionField.java:
```diff
@@ -70,14 +100,14 @@ public boolean equals(Object other) {
 
     PartitionField that = (PartitionField) other;
-    return sourceId == that.sourceId
+    return sourceIds.toString().equals(that.sourceIds.toString())
```
Just wondering why not `Objects.equals(sourceIds, that.sourceIds);` do we need the extra string conversion?

jbonofre replied:
Yeah, either way is fine. Let me do the change if you prefer.

ajantha-bhat commented on .palantir/revapi.yml:
```diff
@@ -1177,6 +1177,10 @@ acceptedBreaks:
       old: "class org.apache.iceberg.Metrics"
       new: "class org.apache.iceberg.Metrics"
       justification: "Java serialization across versions is not guaranteed"
+    - code: "java.class.defaultSerializationChanged"
```
Doesn't this breaks backward compatibility? and this is not allowed for minor versions.

We need to implement custom `writeObject` and `readObject`  to ensure we can serialize the field if it has only `sourceId`?

```
private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
    in.defaultReadObject(); // read all default fields

    // Ensure backward compatibility
    if (sourceIds == null && sourceId != null) {
        sourceIds = Collections.singletonList(sourceId);
    }
}

private void writeObject(ObjectOutputStream out) throws IOException {
    out.defaultWriteObject(); // only write sourceIds
}

```

Note: I know we have handled the JSON serialization properly in `PartitionSpecParser`. Just wondering about default java serializer here.

jbonofre replied:
I think we only use JSON serialization. I don't think we are using Java serialization. Do we ?

ajantha-bhat replied:
Yes, We don't use it. But I don't see a general guidelines about backward compatibility for Java serialization. Let us wait for others opinion.

RussellSpitzer replied:
I think we are fine here, we don't expect a client/server using this code to be on different library versions.  I believe this currently is hit when using Spark because we are going to serialize the table metadata around. For usages like this I don't want to have to support multiple versions on driver / executor.

ajantha-bhat commented on api/src/main/java/org/apache/iceberg/PartitionSpec.java:
```diff
@@ -601,17 +601,26 @@ public Builder alwaysNull(String sourceName) {
-      fields.add(new PartitionField(sourceId, fieldId, name, transform));
+    Builder add(List<Integer> sourceIds, int fieldId, String name, Transform<?, ?> transform) {
+      // we use the first entry in the source-ids list here
+      checkAndAddPartitionName(name, sourceIds.get(0));
```
How does this work for multi arg partition field? 
Do we need a logic that accepts the `sourceIds` and resolve the name of multi arg transform?

jbonofre replied:
For now, I'm using the first element. I can update `checkAndAddPartitionName` to consider all elements in source ids.

jbonofre replied:
OK, let me refactore this part.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionSpecParser.java:
```diff
@@ -98,7 +101,16 @@ static void toJsonFields(UnboundPartitionSpec spec, JsonGenerator generator) thr
+          generator.writeNumber(value);
+        }
+        generator.writeEndArray();
+      }
```
nit: Iceberg follows the new line spacing for code blocks 
https://iceberg.apache.org/contribute/?h=contribut#block-spacing

jbonofre replied:
Ack, fixing.

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/PartitionField.java:
```diff
@@ -19,26 +19,49 @@
+   * @deprecated since 1.10.0, will be removed in 2.0.0; use PartitionField(sourceIds, fieldId,
+   *     name, transform) instead.
+   */
+  @Deprecated
```
I don't actually have a problem with keeping this method? It's essentially just going to save callers doing a List.of(element) and I don't think we are are really adding much complexity?

jbonofre replied:
That's a good point. The reason why I flagged deprecated is to "force" the callers to use source-ids "style" (I don't deprecated in the builder as the current transforms support only one argument).
I'm OK to remove the `@Deprecated` here.

jbonofre replied:
I removed the deprecation.

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/UnboundPartitionSpec.java:
```diff
@@ -54,17 +55,17 @@ private PartitionSpec.Builder copyToBuilder(Schema schema) {
 
     for (UnboundPartitionField field : fields) {
-      Type fieldType = schema.findType(field.sourceId);
+      Type fieldType = schema.findType(field.sourceIds.get(0));
```
I know we don't have a multi-arg transform to resolve here yet but this doesn't seem like the right thing to do. I think Transforms.fromString needs to be modified to accept fromString(list[types], transformName)

jbonofre replied:
Yes, agree. Let me update that.

jbonofre commented on api/src/main/java/org/apache/iceberg/PartitionField.java:
```diff
@@ -19,26 +19,49 @@
+  @Deprecated
   public int sourceId() {
-    return sourceId;
+    Preconditions.checkArgument(sourceIds.size() == 1, "Only one source is supported");
```
Better yeah. Let me update. Thanks for the proposal !

jbonofre replied:
Done.

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/UnboundPartitionSpec.java:
```diff
@@ -118,7 +128,12 @@ public String transformAsString() {
 
     public int sourceId() {
-      return sourceId;
+      Preconditions.checkArgument(sourceIds.size() == 1, "Only one source is expected");
```
cannot call sourceId() on a transform with multiple arguments

RussellSpitzer commented on core/src/test/java/org/apache/iceberg/TestPartitionSpecParser.java:
```diff
@@ -95,6 +96,56 @@ public void testFromJsonWithFieldId() {
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessage(
+            "Cannot parse partition field, either source-id or source-ids has to be present");
+  }
```
We should have some tests for "toJson" as well? 

We also need a check that the validation for identity transforms still holds true. Ie you cannot make a multi-arg transform with the same name as any of the columns

jbonofre replied:
I added a test in `TestPartitionSpecParser` testing when neither `source-id` and `source-ids` are provided: `testFromJsonWithoutSourceIdAndSourceIds()`.
I will add additional tests regarding your previous mention.

RussellSpitzer reviewed (commented):
Overall this looks pretty good to me. I think we are missing some testing but the approach looks correct. There are a few utility methods we need to fix up as well to accept sourceIds

jbonofre commented:
Thanks @RussellSpitzer ! Let me fix the util methods and add tests. Thanks again !

jbonofre commented on core/src/main/java/org/apache/iceberg/PartitionSpecParser.java:
```diff
@@ -132,15 +147,29 @@ private static void buildFromJsonFields(UnboundPartitionSpec.Builder builder, Js
+              + SOURCE_ID
+              + " or "
+              + SOURCE_IDS
+              + " has to be present");
```
`source-id` and `source-ids` are exclusive, meaning that if `source-id` is present, `source-ids` should not be there, and if `source-ids` is present `source-id` should not be there.
I'm adding a check to raise to check if not both are provided.

jbonofre replied:
I added a check to guarantee we only have `source-id` or `source-ids`, not both (and I added a corresponding test for that).

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/PartitionSpec.java:
```diff
@@ -625,7 +634,7 @@ PartitionSpec buildUnchecked() {
   static void checkCompatibility(PartitionSpec spec, Schema schema) {
     for (PartitionField field : spec.fields) {
-      Type sourceType = schema.findType(field.sourceId());
+      Type sourceType = schema.findType(field.sourceIds().get(0));
```
We should be extending this logic as well to handle multi arg sources

jbonofre replied:
Yup. I'm doing a new update.

jbonofre replied:
I introduced new `canTransform(List)` and `getResultType(List)` to support multi-args in transforms.
I didn't remove the "old" methods, and I propose to deprecate it.
Thoughts ?

jbonofre commented:
Hey guys. I was traveling this week. I'm now back on this pr, updating according to the comments.

jbonofre commented:
I did a first update to introduce multi-args in `Transform`. I will check/update the tests too.

pvary commented on api/src/main/java/org/apache/iceberg/PartitionSpec.java:
```diff
@@ -627,7 +636,10 @@ PartitionSpec buildUnchecked() {
+      List<Type> sourceTypes = Lists.newArrayList();
+      for (int sourceId : field.sourceIds()) {
+        sourceTypes.add(schema.findType(sourceId));
+      }
```
nit: newline

pvary commented on api/src/main/java/org/apache/iceberg/PartitionSpec.java:
```diff
@@ -602,17 +602,26 @@ public Builder alwaysNull(String sourceName) {
-    Builder add(int sourceId, int fieldId, String name, Transform<?, ?> transform) {
-      checkAndAddPartitionName(name, sourceId);
-      fields.add(new PartitionField(sourceId, fieldId, name, transform));
+    Builder add(List<Integer> sourceIds, int fieldId, String name, Transform<?, ?> transform) {
```
Do we need a preconditions against null sourceIds in the list?

pvary commented on api/src/main/java/org/apache/iceberg/UnboundPartitionSpec.java:
```diff
@@ -88,14 +89,23 @@ Builder withSpecId(int newSpecId) {
     }
 
+    Builder addField(
+        String transformAsString, List<Integer> sourceIds, int partitionId, String name) {
```
null checks?

jbonofre replied:
Same comment: let me add a null element check.

pvary commented on api/src/main/java/org/apache/iceberg/PartitionSpec.java:
```diff
@@ -636,15 +648,20 @@ static void checkCompatibility(PartitionSpec spec, Schema schema) {
+              "Cannot partition by non-primitive source field: %s (%s)",
+              field,
+              sourceType);
+        }
```
nit: newline

jbonofre commented:
Here we go again 😄

jbonofre commented:
I'm back (again 😄 ) on Iceberg, resuming my different PRs.

jbonofre committed:
Core: add source-ids support for multi-args transform

jbonofre committed:
Core: accept revapi breaks for PartitionField and SortField serialization changes

jbonofre committed:
Core: fix spotless formatting

sfc-gh-rspitzer commented on api/src/main/java/org/apache/iceberg/PartitionField.java:
```diff
@@ -19,26 +19,42 @@
   /** Returns the field id of the source field in the {@link PartitionSpec spec's} table schema. */
   public int sourceId() {
-    return sourceId;
+    return sourceIds.get(0);
```
I still think there should be a precondition that sourceIds length is = 1,  just to make sure folk are using this correctly

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/PartitionField.java:
```diff
@@ -58,7 +74,10 @@ public String name() {
+    if (sourceIds.size() == 1) {
+      return fieldId + ": " + name + ": " + transform + "(" + sourceIds.get(0) + ")";
+    }
+    return fieldId + ": " + name + ": " + transform + "(" + sourceIds + ")";
```
The multi-source-id branch uses `List.toString()` directly, which produces square brackets in the output: `transform([1, 2, 3])` instead of `transform(1, 2, 3)`.

Using a `Joiner` (as done elsewhere in the api module, e.g. `UnboundPredicate`, `BoundSetPredicate`) would fix this and also eliminate the need for the `size() == 1` special case:

```java
private static final Joiner COMMA = Joiner.on(", ");

@Override
public String toString() {
  return fieldId + ": " + name + ": " + transform + "(" + COMMA.join(sourceIds) + ")";
}
```

Same applies to `SortField.toString()`.

jbonofre committed:
Core: harden PartitionField and PartitionSpec for multi-arg transforms

Add a precondition guard on PartitionField.sourceId() to prevent misuse
with multi-arg transforms, fix toString() to use Joiner instead of
List.toString(), migrate all sourceId() call sites in PartitionSpec to
use sourceIds(), and add a consistency check in PartitionSpecParser when
both source-id and source-ids are present.

jbonofre committed:
Core: use LOG.warn instead of precondition for conflicting source-id/source-ids

Follow reviewer guidance to warn rather than throw when parsing, since
Iceberg generally avoids hard failures during deserialization.

jbonofre committed:
Core: fix test calling sourceId() on multi-arg partition field

The testFromJsonFieldsWithMultipleSourceIds test was calling
sourceId() on a field with multiple source IDs, which now throws
after the hardening in 7a4cc928b. Use sourceIds() instead.

jbonofre committed:
Trigger CI re-run

jbonofre committed:
Trigger CI re-run

moomindani commented:
Hi @jbonofre, are you still planning to pursue this PR? I'm asking because I recently implemented the v3 read-tolerance side of this in PyIceberg (apache/iceberg-python#3630: fields with multiple `source-ids` parse and are treated as unknown transforms, with spec-compliant `source-id`/`source-ids` round-trip), and I'd like to help bring the Java implementation to that baseline — today, metadata containing a `source-ids` field cannot be read at all. Happy to help here if that's useful.
